File: 0668-kth-smallest-number-in-multiplication-table/0668-kth-smallest-number-in-multiplication-table.java
class Solution {
    public int fun(int m,int n,int guess){
        int count=0; 

        for(int i=1;i<=m;i++){
            count=count+Math.min(n,guess/i);
        }
        return count;
    }

    public int findKthNumber(int m, int n, int k) {
        int low=1;
        int high=m*n; 
        int res=-1;

        while(low<=high){
            int guess=(low+high)/2;

            int ans=fun(m,n,guess);

            if(ans<k){
                low=guess+1;
            }
            else{
                res=guess;
                high=guess-1;
            }
        }
        return res;
    }
}

File: 0668-kth-smallest-number-in-multiplication-table/0668-kth-smallest-number-in-multiplication-table.go
func fun(m int, n int, guess int) int {
    count:=0

    for i:=1;i<=m;i++{
        count=count+min(n,guess/i)
    }
    return count
}

func findKthNumber(m int, n int, k int) int {
    low:=1
    high:=m*n
    res:=-1

    for low<=high{
        guess:=(low+high)/2

        ans:=fun(m,n,guess)

        if ans<k{
            low=guess+1
        }else{
            res=guess
            high=guess-1
        }
    }
    return res
}

func min(a int, b int) int {
    if a<b{
        return a
    }
    return b
}